Add centering by longest line

cappe987 opened the pull request:
Centering by longest line requires inputting all lines before starting output, to determine which is longest. Use a tail queue for this to insert at end and then iterate from start. The lines are freed after being output.

Unlike the static read buffer, we can't reuse the list buffers in an easy way if we are iterating over multiple input files. So they need to be freed to avoid memory leaks.

Mango0x45 commented:
Thanks for the PR!  I'll review it later today when I have the time.

Mango0x45 commented on center.c:
```diff
@@ -88,7 +101,10 @@ main(int argc, char **argv)
 				rval = EXIT_FAILURE;
 			} else {
-				center(fp);
+				if(use_longest)
```
You're missing a space between `if` and `(use_longest)`

cappe987 replied:
Ah, missed that one. Fixed!

cappe987 commented:
> ```
> This is one sentence.  This is another.
>                      ^^
> ```
> 
> This makes reading the output of `git log` a bit easier since it'll be using a monospace font.

I did this now. But this is not something I've heard of before. My `git log` is and has always been monospace only. Could it depend on the font set for the terminal?

cappe987 commented:
Fixed the missing space in the code and added extra spaces in the commit. Is using two spaces just personal preference, or is this something a lot of people do? I've never heard of it before.

Mango0x45 commented:
> Is using two spaces just personal preference, or is this something a lot of people do?

For text consumed using a monospace font, it's quite common to see double spacing.  Over time there seems to be less and less people doing it (because they just aren't aware of it I suppose), but it is still pretty common to see.  The easiest example is to just read any manpage on your system with `man` and see how it automatically double spaces sentences.

Mango0x45 commented:
From what I've tested, it seems to work for the most part.  There is one issue
though which is that in the cases where `center()` is called on `stdin` (so when
no command-line arguments are provided, or when the filename '-' is provided) it
calls the regular `center()` funciton instead of `center_by_longest()`.  This is
easily fixed with a couple added `if` statements, but I think instead of adding
all these conditional checks everywhere maybe we should use a function pointer
for `center()`/`center_by_longest()` just as we do for `utf8len()`/`noesclen()`.

Mango0x45 commented on center.c:
```diff
@@ -124,6 +140,63 @@ center(FILE *fp)
+void
+center_by_longest(FILE *fp)
+{
+	static char *buffer = NULL;
+	static size_t bs = 0;
```
There's no reason to make these variables `static`, as the `getline()` loop sets both to `NULL` and `0` at the end of each iteration so we aren't saving any allocations anyways.

Mango0x45 commented on center.c:
```diff
@@ -124,6 +140,63 @@ center(FILE *fp)
+void
+center_by_longest(FILE *fp)
+{
+	static char *buffer = NULL;
+	static size_t bs = 0;
```
There's no reason to make these variables static, as the getline() loop sets both to NULL and 0 at the end of each iteration so we aren't saving any allocations anyways.

cappe987 replied:
Actually, there is. `man getline`
> This buffer should be freed by the user program even if getline() failed.

When it has read the last line and tries to read another it will fail (because no more input), but will have allocated a buffer anyways. Because it doesn't run the loop it still has the buffer that was never used. It is in the same situation as `center()` where it has one allocated buffer that is never freed. Meaning, `center_by_longest()` also has a (intentional) memory leak. If it was not static we would leak that final one allocation.

Mango0x45 replied:
Huh, I looked at the manual while writing this comment for that specific line and couldn't find it, lol.  Should be fine then!

cappe987 commented:
> From what I've tested, it seems to work for the most part. There is one issue though which is that in the cases where `center()` is called on `stdin` (so when no command-line arguments are provided, or when the filename '-' is provided) it calls the regular `center()` funciton instead of `center_by_longest()`. This is easily fixed with a couple added `if` statements, but I think instead of adding all these conditional checks everywhere maybe we should use a function pointer for `center()`/`center_by_longest()` just as we do for `utf8len()`/`noesclen()`.

Sorry, forgot to test that. Good idea with function pointer. Changed it to that.

Mango0x45 commented on center.c:
```diff
@@ -44,6 +53,7 @@ extern char *optarg;
 int rval;
 long width;
 int (*lenfunc)(const char *) = noesclen;
+void (*centerfunc)(FILE *) = center;
```
This isn't really used anywhere except for `main()`, so we can probably make it a local variable to `main()` instead of a global.

cappe987 replied:
:+1:
